Add LLVM level allocation optimization pass

yuyichao opened the pull request:
Test might be added later.

Just to give a taste of it though

```julia
julia> function test_opt(a::Float64)
           s = Ref{Float64}()
           c = Ref{Float64}()
           ccall((:sincos, Base.libm_name), Void, (Float64, Ptr{Float64}, Ptr{Float64}), a, s, c)
           s[], c[]
       end
test_opt (generic function with 1 method)

julia> for i in 1
           @time test_opt(1.0)
       end
  0.000005 seconds

julia> @code_llvm test_opt(1.0)
```
```llvm
define void @julia_test_opt_63022([2 x double] addrspace(11)* noalias nocapture sret, double) #0 !dbg !5 {
top:
  %2 = alloca i64, align 8
  %3 = bitcast i64* %2 to %jl_value_t*
  %4 = alloca [2 x double], align 8
  %5 = bitcast %jl_value_t* %3 to double*
  %6 = getelementptr inbounds [2 x double], [2 x double]* %4, i64 0, i64 0
  call void inttoptr (i64 140684203667904 to void (double, double*, double*)*)(double %1, double* %6, double* %5)
  %7 = load i64, i64* %2, align 16
  %8 = getelementptr inbounds [2 x double], [2 x double]* %4, i64 0, i64 1
  %9 = bitcast double* %8 to i64*
  store i64 %7, i64* %9, align 8
  %10 = bitcast [2 x double]* %4 to i8*
  %11 = bitcast [2 x double] addrspace(11)* %0 to i8 addrspace(11)*
  call void @llvm.memcpy.p11i8.p0i8.i32(i8 addrspace(11)* %11, i8* %10, i32 16, i32 8, i1 false)
  ret void
}
```
```julia
julia> function f(a)
           b = Ref(a)
           b[] += 1
           b[] -= 1
           b[] += 2
           b[]
       end
f (generic function with 1 method)

julia> @code_llvm f(2)
```
```llvm
define i64 @julia_f_63068(i64) #0 !dbg !5 {
top:
  %1 = add i64 %0, 2
  ret i64 %1
}
```

This takes advantage of the LLVM optimizations to get more precise esape info but it doesn't replace allocation elimination in typeinf, which can also split allocations and do more fancy transformations.

The code generated is not 100% optimum (issue from LLVM optimization order) though that might be better when we can run this later in the pipeline. The current placement for 5.0 is arbitrary....

@nanosoldier `runbenchmarks(ALL, vs = ":master")`

yuyichao committed:
Add LLVM level allocation optimization pass

This can obtain escape information with much higher precision than what we can currently do
in typeinf. However, it does not replace the alloc_elim_pass! in type inference either since
this cannot handle objects with reference fields.

Fix #20452

nanosoldier commented:
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/22684):
```
NanosoldierError: failed to run benchmarks against primary commit: failed process: Process(`make -j3`, ProcessExited(2)) [2]
```
Logs and partial data can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/c18d6a99667aa5183c4d88d6ba4e1093969eff4d/825bfde_vs_bf83397/report.md)
cc @jrevels

yuyichao commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/22684#issuecomment-312994011):
```
NanosoldierError: failed to run benchmarks against primary commit: failed process: Process(`/home/nanosoldier/workdir/tmpSVWZ30/julia -e Pkg.update()`, ProcessSignaled(11)) [0]
```
Logs and partial data can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/de25fac12d2b5c6341cdd7ad1a52ef29ccf01b99/6b62cff_vs_bf83397/report.md)
cc @jrevels

yuyichao commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/22684#issuecomment-313002056):
```
NanosoldierError: failed to run benchmarks against primary commit: failed process: Process(`/home/nanosoldier/workdir/tmpSzyGHq/julia -e Pkg.update()`, ProcessExited(1)) [1]
```
Logs and partial data can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/2f39637b7f440e3b9bf3774df392426a6a1a450b/ee76106_vs_bf83397/report.md)
cc @jrevels

yuyichao committed:
Use a local buffer in `modf`

Fix #21591

yuyichao commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

Keno commented:
Stealing my thunder, eh? That's ok :). I'll try to review this tomorrow.

tkelman commented on src/llvm-alloc-opt.cpp:
```diff
@@ -0,0 +1,385 @@
+    if (auto call = dyn_cast<CallInst>(I)) {
+        if (ptr_from_objref && ptr_from_objref == call->getCalledFunction())
+            return true;
+        // Only use in argument counts, uses in operand bundle doesn't since it cannot escape.
```
doesn't what?

yuyichao replied:
Doesn't count.

tkelman replied:
ah then "uses ... don't" or "use ... doesn't"

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/22684#issuecomment-313144842) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/b942172790f82e58f2497f01e1d6fbdfec6a48a6/6c00ebb_vs_bf83397/report.md). cc @jrevels

tkelman commented on src/llvm-alloc-opt.cpp:
```diff
@@ -0,0 +1,385 @@
+        bool ignore_tag = true;
+        auto orig = it.first;
+        if (optimize && checkUses(orig, 0, ignore_tag)) {
+            // The allocation does not escape or be used in a phi node so none of the derived
```
or get used ?

tkelman commented:
Does this mean we can finally deprecate `&` in ccall?

yuyichao commented:
> Stealing my thunder, eh? That's ok :)

FWIW, the main part of this is just ~4hrs of work after I suddenly got interested in checking if the gcframe placement pass is late enough for us to do this. So nothing much would be lost if you already have a better version working ;-p (Plus I've been implementing Jameson's ideas all along  :trollface: )

> Does this mean we can finally deprecate `&` in ccall?

Yes, that should be the case this PR is the best at doing. There are still rare cases where the `&` in `ccall` can be used to reduce allocation but that only happen with a custom `ptr_arg_cconvert` and `ptr_arg_unsafe_convert` and I don't think anyone is doing it. I think master will be open for PR's replacing `&`s with `Ref`s after this is merged and we can deprecate `&` after making sure nothing goes wrong.

maleadt commented on src/cgutils.cpp:
```diff
@@ -2156,25 +2156,10 @@ static Value *emit_allocobj(jl_codectx_t &ctx, size_t static_size, Value *jt)
 {
     JL_FEAT_REQUIRE(ctx, dynamic_alloc);
     JL_FEAT_REQUIRE(ctx, runtime);
```
Should probably postpone these checks to the actual lowering (but we don't have a `cgctx` there).

yuyichao replied:
That can be passed as a paramter. What would be the correct behavior if the test fail though?

maleadt replied:
abort compilation. currently `JL_FEAT_REQUIRE` just calls `jl_error`, and we use the `cgctx` for current function name & line number, so we'd need to do something else anyway (use DebugLoc, I assume). better leave that to a different PR, for now maybe just:

```
if (!JL_FEAT_TEST(ctx, dynamic_alloc)) jl_error(...)
```

yuyichao replied:
I mean, I don't think we can throw an error in the llvm pass.

maleadt replied:
Hmm I hadn't considered that, as it worked with `_dump_function` but seems to mess up the JIT indeed. Then again, I only use that feature through `_dump_function`...
Maybe a "used features mask" to be checked after lowering? Let's just leave it at what it is now and put that in a different PR.

Keno commented on src/codegen.cpp:
```diff
@@ -6631,24 +6630,14 @@ static void init_julia_llvm_env(Module *m)
+    gc_alloc_args.push_back(T_prjlvalue);
+    jl_alloc_obj_func = Function::Create(FunctionType::get(T_prjlvalue, gc_alloc_args, false),
+                                         Function::ExternalLinkage,
+                                         "julia.gc_alloc_obj");
```
Ideally, this function should get a `noalias` attribute.

Keno commented:
Couple of suggestions:
1. Use LLVM's `PtrUseVisitor` to avoid excessive stack growth in the recursive approach
2. You can call mem2reg as a cleanup step: http://llvm.org/doxygen/namespacellvm.html#a033a44177ba94b77622aae61ff4fb4b2

yuyichao committed:
Improving ieee754_rem_pio2

Fix #22004

yuyichao commented:
1. `noalias` attribute added.
2. I can't figure how to use `PtrUseVisitor` to handle `AddrSpaceCastInst` and it also seems hard to use it to handle the mutation recursion so I decided to use the pattern I'm more familiar with and used a manual stack similar to the one in the GC.
3. `mem2reg` actually doesn't handle this case (at all). `instcombine` and `sroa` does and the new placement for 5.0 takes advantage of that.
4. Added llvm lifetime intrinsics to reuse stack space. Not sure if there's a LLVM helper function that can make doing that easier...

yuyichao commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/22684#issuecomment-313867043):
```
NanosoldierError: failed to run benchmarks against primary commit: failed process: Process(`make -j3`, ProcessExited(2)) [2]
```
Logs and partial data can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/daece5ccae64223bff867186d044ec2a326c09e2/c176fc8_vs_0c9a858/report.md)
cc @jrevels

yuyichao commented:
The issue in the lifetime computation should be fixed...

@nanosoldier runbenchmarks(ALL, vs = ":master")

yuyichao commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/22684#issuecomment-313870411) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/a376a228adcfa4a551d23cc7062709745bea25db/c7f3ce2_vs_0c9a858/report.md). cc @jrevels

tkelman commented on src/llvm-alloc-opt.cpp:
```diff
@@ -0,0 +1,733 @@
+        if (!bbs.insert(bb).second)
+            continue;
+        if (I == &*bb->begin()) {
+            // The there's no use in or after this bb. If this bb is not dominated by
```
redundant "The" ?

vtjnash commented:
Need to make sure that for:
```julia
julia> f(fptr, x) = ccall(fptr, Void, (Ref{Integer},), x)
f (generic function with 1 method)

julia> @code_llvm f(C_NULL, 0)
```
We are still emitting the gc-root for `x` even when we elide the allocation of the `Ref` that we used to declare it.

yuyichao commented on base/math.jl:
```diff
@@ -754,9 +754,9 @@ function ieee754_rem_pio2(x::Float64)
+    y = Ref{NTuple{2,Float64}}()
+    n = ccall((:__ieee754_rem_pio2, openspecfun), Cint, (Float64, Ptr{Void}), x, y)
+    return (n, y[])
 end
```
The ordering of the pass on 3.9 and 4.0 still make it slightly less efficient than the llvmcall version.

yuyichao commented on src/cgutils.cpp:
```diff
@@ -2149,25 +2149,10 @@ static Value *emit_allocobj(jl_codectx_t &ctx, size_t static_size, Value *jt)
-    return v;
+    return ctx.builder.CreateCall(prepare_call(jl_alloc_obj_func),
+                                  {ptls_ptr, ConstantInt::get(T_size, static_size),
+                                          maybe_decay_untracked(jt)});
```
Does this have to be marked on the callsite?

yuyichao commented on src/llvm-alloc-opt.cpp:
```diff
@@ -0,0 +1,733 @@
+// This file is a part of Julia. License is MIT: https://julialang.org/license
+
+#define DEBUG_TYPE "alloc_opt"
+#undef DEBUG
```
.... No idea....

yuyichao commented on src/llvm-alloc-opt.cpp:
```diff
@@ -0,0 +1,733 @@
+        Instruction *orig_i = cur->orig_i;
+        Instruction *new_i = cur->new_i;
+        if (isa<LoadInst>(user) || isa<StoreInst>(user)) {
+            user->replaceUsesOfWith(orig_i, new_i);
```
The type of `orig_i` and `new_i` are always the same ignoring addrspace. If they are not, they only differ in addrspace.

yuyichao commented on src/llvm-alloc-opt.cpp:
```diff
@@ -0,0 +1,733 @@
+                new_gep = GetElementPtrInst::Create(gep->getSourceElementType(),
+                                                    new_i, IdxOperands,
+                                                    gep->getName(), gep);
+            }
```
`copyMetadata` does that.

yuyichao commented on src/llvm-alloc-opt.cpp:
```diff
@@ -0,0 +1,733 @@
+    const auto &dbg = I->getDebugLoc();
+    auto derived = new AddrSpaceCastInst(newI, T_pjlvalue_der, "", I);
+    derived->setDebugLoc(dbg);
+    auto cast = new BitCastInst(derived, T_ppjlvalue_der, "", I);
```
AFAICT it's still pretty pedantic....

yuyichao commented on src/llvm-alloc-opt.cpp:
```diff
@@ -0,0 +1,733 @@
+                cast2->setDebugLoc(dbg);
+                auto tagaddr = GetElementPtrInst::Create(T_pjlvalue, cast,
+                                                         {ConstantInt::get(T_size, -1)},
+                                                         "", orig);
```
Yeah, It was tested but not after making it conditional....... I'm actually not sure how to actually trigger this....

yuyichao commented:
> `Ref{Integer}`

Being exposed by, fixed in and tested in https://github.com/JuliaLang/julia/pull/22734

yuyichao commented on src/llvm-alloc-opt.cpp:
```diff
@@ -0,0 +1,733 @@
+    if (I->user_empty())
+        return true;
+    CheckInstFrame cur{I, 0, I->user_begin(), I->user_end()};
+    stack.clear();
```
This one might actually not be empty due to the short circuit return. (The other two should.)

tkelman commented on src/llvm-alloc-opt.cpp:
```diff
@@ -0,0 +1,746 @@
+/**
+ * Promote `julia.gc_alloc_obj` which do not have escaping root to a alloca and
+ * lower other ones to real GC allocation.
+ * Uses that are not considerer to escape the object (i.e. heap address) includes,
```
considered ?

tkelman commented on src/llvm-alloc-opt.cpp:
```diff
@@ -0,0 +1,746 @@
+ *
+ *     The results of these cast instructions will be scanned recursively.
+ *
+ * All other uses are considered escape conservatively.
```
considered to escape ?

tkelman commented on src/llvm-alloc-opt.cpp:
```diff
@@ -0,0 +1,746 @@
+    // Makes sure that the end is inserted before nearby start.
+    // We insert start before the allocation call, if it is the first safepoint we find for
+    // another instruction, it's better if we insert the end before the start instead of the
+    // allocation so that the two allocation do not have overlapping lifetime.
```
so that the two allocations ?

Keno commented on src/llvm-alloc-opt.cpp:
```diff
@@ -0,0 +1,746 @@
+            // `memcmp` by coverting to our own intrinsic and lower it after the gc root pass.
+            if (ptr_from_objref && ptr_from_objref == call->getCalledFunction())
+                return true;
+            // Only use in argument counts, use in operand bundle doesn't since it cannot escape.
```
Only skip the "jl_roots" arg bundle? In general arg bundles escape in unknown ways per the spec.

yuyichao replied:
It is only skipping `jl_roots` arg bundle. I'll update the comment.

Keno commented:
I'd like to see some LLVM level tests. Also, why introduce a new allocation intrinsic, rather than operating on the actual GC functions? It seems like it may be useful to put this pass in the pipeline twice (e.g. if we want to do LLVM level inlining).

yuyichao commented:
> why introduce a new allocation intrinsic

1. The `jl_gc_pool_alloc` does not provide the actually needed size (`osize` contains tag and is rounded up by a lot, especially for small objects like `Int8`, `Float32`)
2. It avoids introducing the tag store in the IR too early. That is much harder to optimize out.

I've also thought about running this twice. It should be easy to do by disabling the intrinsic lowering with a flag.

yuyichao commented:
All actionable review comments addressed.

Keno commented:
Would it make sense to do the lowering from intrinsics to actual call as part of GC root placement? That pass runs only once anyway and it doesn't make sense to do any allocation optimization afterward.

Keno commented on test/llvmpasses/alloc-opt.jl:
```diff
@@ -0,0 +1,204 @@
+
+## WIP
+
+isz = sizeof(UInt) == 8 ? "i64" : "i32"
```
We should really not have the pass depend on the how architecture (and have it look at the target arch/triple instead).

yuyichao replied:
While the integer size is easy to fix, its actually a pretty complex logic. The pass relies on details on object layout which not only depend on 32/64bit but also the architecture and even the OS. In particular, the pool offset depends on other members in the tls buffer including the size of `jl_jmp_buf`.

P.S. just notice that the `WIP` should be removed.....

yuyichao commented:
> as part of GC root placement

I think it makes less sense than doing it here. This pass already need to collect all the allocations and the last invokation of this pass might as well lower the remaining ones.

Keno commented:
Why is the lowering not relatively trivial?

Keno commented:
Or if it is, GC root lowering looks at all the calls and rewrites most of them anyway (because of the cc convention), so it seems like a fine place to do it.

Keno commented:
The concern is that this pass might not know whether it's the last one (in IPO pipelines the pass manager will automatically rerun parts of the pipeline).

yuyichao commented:
Moved the intrinsic lowering to gcframe lowering pass and updated the test.

yuyichao commented:
Travis failure looks unrelated and is happenning everywhere....

yuyichao commented:
More comments?

Keno commented:
I'll try to take another pass through this tonight or tomorrow morning.

Keno commented on src/llvm-alloc-opt.cpp:
```diff
@@ -0,0 +1,667 @@
+    }
+}
+
+bool AllocOpt::isSafepoint(Instruction *inst)
```
Pull this out into a helper for both this and the GC lowering code?

yuyichao replied:
Does the lowering pass actually need this function? It has similar logic but does different things for different branches.

Keno replied:
It doesn't need this function, but I'd like to at least share the "Known functions emitted in codegen that are not safepoints" part, so we don't have to make changes there in multiple places.

yuyichao replied:
Also note that when the lowering pass recognize more functions as non safepoint, we don't necessarily want to update this pass to include those. Here I included the same list just to be safe. In principle, this needs a list of functions that codegen assums to be not safepoint. There can be functions that are never safepoint but as long as neither codegen or llvm can insert a call into a unsafe use chain it doesn't need to be treated as non safepoint here.

Keno commented on src/llvm-alloc-opt.cpp:
```diff
@@ -0,0 +1,667 @@
+{
+    if (!alloc_obj)
+        return false;
+    std::map<CallInst*,size_t> allocs;
```
std::map is generally discouraged in LLVM passes because of nondeterministic iteration order. Can you use a data structure with deterministic iteration?

yuyichao replied:
Seems that I actually don't need to lookup anything in it so I'll just use a vector instead...

Keno commented on src/llvm-alloc-opt.cpp:
```diff
@@ -0,0 +1,667 @@
+            if (ptr_from_objref && ptr_from_objref == call->getCalledFunction())
+                return true;
+            auto opno = use->getOperandNo();
+            // Uses in `jl_roots` operand bundle are not counted as escaping, everything else do.
```
Nit: `everything else is`.

Keno commented on src/llvm-alloc-opt.cpp:
```diff
@@ -0,0 +1,667 @@
+    };
+    // Both `orig_i` and `new_i` should be pointer of the same type
+    // but possibly different address spaces. `new_i` is always in addrspace 0.
+    auto replace_inst = [&] (Instruction *user) {
```
You can technically mutate instructions in place, but I'm fine with this implementation as well.

yuyichao commented:
Updated comment and switched to a vector instead for recording the allocations.
I did not pull the `isSafePoint` out since there currently isn't a use in the lowering pass. We can certainly split it out when there is one.

yuyichao commented:
I'd like to not make any other changes about the organization, especially since I still don't think it's a good idea to move the allocation that late in the pipeline. It disables the llvm constant folding of write barrier on 5.0....

There are two FreeBSD timeout and I have no way to debug it so I'll just wait for someone else to figure out what's wrong there or fix it or merge this PR.

yuyichao commented:
And for anyone who want to use this on master, the hanging test on FreeBSD is `file`.

iblis17 commented:
the `distributed` testsuit absent in log also.
FreeBSD CI keep hanging randomly ... It seems start from 5ea8c7c8f6
I guess that the test case add in #22566 another bug...

yuyichao commented:
> the distributed testsuit absent in log also.
JULIA_TEST_MAX_RSS is set so it is moved to worker one and will only run if everything else finishes.

iblis17 commented:
👌 
In case of hanging `file`, feel free to rerun CI.
There is a `rebuild` button at the top-right corner of buildbot UI.

yuyichao commented:
[Restarted one](https://julia.iblis.cnmc.tw/#/builders/1/builds/577) passed so I'm merging this now....

iblis17 commented:
got some compliation warnning from master e1a604e088
```cpp
/usr/home/iblis/git/julia/src/llvm-alloc-opt.cpp:639:53: warning: braces around scalar initializer [-Wbraced-scalar-init]                         
                                                    {ConstantInt::get(T_size, -1)});                                                              
                                                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                                                                
/usr/home/iblis/git/julia/src/llvm-late-gc-lowering.cpp:1173:50: warning: braces around scalar initializer [-Wbraced-scalar-init]                 
                                                 {ConstantInt::get(T_size, -1)});                                                                 
                                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                                                                   
1 warning generated.                                                                                                                              
1 warning generated.                                                                                                                              
```

JeffBezanson commented:
This is awesome!

From the nanosoldier run, it looks like we don't have any benchmarks that meaningfully improve with this? We should try to add some.

yuyichao commented:
Regression on this will probably show up in future benchmarks as we have more and more use of `Ref` in `ccall`s.

We are in general very good at avoiding performance pitfails in the benchmarks....
